### ISSUE TITLE
Interpretation of Yolo10's and Yolo11's output using OpenVINO

### ISSUE THREAD

**Issue by nikpayam** [ultralytics/ultralytics#18754]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

# Question
I have converted Yolo10 and Yolo11 nano models to OpenVINO model. When I use `YOLO` function and pass these models to detect objects in an image, the output is understandable. But, if I use OpenVINO's API to infer the model, the output is weird.
Consider the following image:
![Image](https://github.com/user-attachments/assets/35df0fe1-d190-4efd-ad15-063ef75d3d52)

The Yolo10n model finds five cars and Yolo11 finds five cars and two trucks in the image, but the shape of the output is (1,300,6) for Yolo10 and (1,84,8400) for Yolo11.
**First:** Does YOLO10 need nms?

Moreover, the output data is not meaningful. Consider Yolo10: I guess the third dimension in Yolo10's output should correspond to `(x, y, w, h, confident score, class number)`, but the output data shows something else: 
* The 5th column (which should corresponds to confident score) is always 1.0.
* The 6th column (which should corresponds to class number) contains 41,67 and 73 (not 2).

Note that car class number is 2 in pretrained Yolo models. 

**Second:** Could someone please explain the structure of the output of Yolo10 model when I use OpenVINO's API.

# Code for converting Yolo model to OpenVINO
```
from ultralytics import YOLO

#model = YOLO("yolov10n.pt")
model = YOLO("yolo11n.pt")

result = model.export(format="openvino", device='cpu', dynamic=False, half=False, imgsz=(640,640))
```

# Code for inferring images using Yolo classes with OpenVINO models
```
from ultralytics import YOLO

#model = YOLO("./yolov10n_openvino_model/")
model = YOLO("./yolo11n_openvino_model/")

source = "../car.jpg"

result = model.predict(source, show=True, show_labels=True, show_conf=True, save=True, imgsz=(640,640), conf=0.2, device='cpu')
```

# Code for inferring images using OpenVINO api
```
import cv2
import numpy as np
import openvino as ov

core = ov.Core()

model_path = "./yolov10n_openvino_model/yolov10n.xml"
#model_path = "./yolo11n_openvino_model/yolo11n.xml"

image_path = "../car.jpg"
device_name = "CPU"

# ------------- read model -------------

model = core.read_model(model_path)

# ------------- setup input -------------

# Read input image
image = cv2.imread(image_path)

# Add N dimension
input_tensor = np.expand_dims(image, 0)

# ------------- apply preprocessing -------------

ppp = ov.preprocess.PrePostProcessor(model)

_, h, w, _ = input_tensor.shape

# 1) Set input tensor information:
# - input() provides information about a single model input
# - reuse precision and shape from already available `input_tensor`
# - layout of data is 'NHWC'
ppp.input().tensor() \
    .set_shape(input_tensor.shape) \
    .set_element_type(ov.Type.u8) \
    .set_layout(ov.Layout('NHWC'))  # noqa: ECE001, N400

# 2) Adding explicit preprocessing steps:
# - apply linear resize from tensor spatial dims to model spatial dims
ppp.input().preprocess().resize(ov.preprocess.ResizeAlgorithm.RESIZE_LINEAR)

# 3) Here we suppose model has 'NCHW' layout for input
ppp.input().model().set_layout(ov.Layout('NCHW'))

# 4) Set output tensor information:
# - precision of tensor is supposed to be 'f32'
ppp.output().tensor().set_element_type(ov.Type.f32)

# 5) Apply preprocessing modifying the original 'model'
model = ppp.build()

# ------------- Load model to device -------------

compiled_model = core.compile_model(model, device_name) #load model with preprocessing

# ------------- Create infer request and do inference synchronously -------------

results = compiled_model.infer_new_request({0: input_tensor})

# ------------- Process output -------------

predictions = results[0]
print(predictions)
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @nikpayam, thank you for your detailed question and for your interest in Ultralytics 🚀! We appreciate you sharing the code, details, and context for your inquiry.

We recommend starting by immersing yourself in the [Docs](https://docs.ultralytics.com) 📖, where you can find detailed guides and tutorials, including [Python usage](https://docs.ultralytics.com/usage/python/) and [CLI usage](https://docs.ultralytics.com/usage/cli/), which may help clarify your questions about model outputs and usage.

If this pertains to a 🐛 Bug with unexpected model output when using OpenVINO, we kindly ask for a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to further aid debugging. For debugging purposes, it would also help to understand the exact version of Ultralytics, Python, and OpenVINO being used, alongside the pre-trained weights (`yolov10n.pt`/`yolo11n.pt`) involved.

If this is a ❓ Question about custom workflows or integration, providing as much detailed information as possible gives us and the community more room to assist. For instance, this could include clarifications about how you are interpreting the output dimensions `(1, 300, 6)` and `(1, 84, 8400)`, and how they tie back to OpenVINO's inference behavior and postprocessing constraints.

In the meantime, please ensure that:
- You have the latest version of the `ultralytics` package installed. To upgrade, use:
```bash
pip install -U ultralytics
```
- You verify your OpenVINO-exported models are operating consistently with the [OpenVINO documentation](https://docs.ultralytics.com/modes/export/#openvino) and correspond to the expected postprocessing behavior described in YOLO (e.g., confidence score scaling, NMS, etc.).

Join the Ultralytics community to exchange ideas and seek further help:
- For real-time discussions, hop onto [Discord](https://discord.com/invite/ultralytics) 🎧.
- Participate in meaningful threaded discussions on [Discourse](https://community.ultralytics.com).
- Share your findings or questions on Reddit via the [Subreddit](https://reddit.com/r/Ultralytics).

## Verified Environments

YOLO models and exports, including OpenVINO, operate in a variety of verified environments. To quickly test or debug your setup, you might consider these GPU-enabled environments:
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- For cloud VM setups, check [GCP](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/), or [Docker](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/).

## Continuous Integration

If all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing as indicated by this badge: <a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>, it verifies model export and inference correctness for various environments and configurations.

Rest assured, this is an automated response 🤖 to help streamline initial support, but a human Ultralytics engineer will be with you shortly to assist further. Thank you for your patience and for being part of the Ultralytics community! 🚀

**Comment by Y-T-G** [comment]
YOLOv10 doesn't need NMS. The output is `(x1, y1, x2, y2, score, label)`

**Comment by nikpayam** [comment]
I thought it should be like this too. But as I said, if we use `YOLO` function, it detects only 5 cars, but the output of OpenVINO API has 300 rows! Also, all 300 detected objects have a score of 1 and their label does not have the number 2, which belongs to the car. I really got confused.

**Comment by Y-T-G** [comment]
Probably something wrong with your preprocessing

**Comment by nikpayam** [comment]
I wrote the codes that produce these results. Could you take a look at them?

**Comment by Y-T-G** [comment]
It's not correct. The obvious issues are wrong resizing method, missing normalization.

There's a C++ example for OpenVINO.

https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-OpenVINO-CPP-Inference/inference.cc

**Comment by nikpayam** [comment]
> It's not correct. The obvious issues are wrong resizing method, missing normalization.
> 
> There's a C++ example for OpenVINO.
> 
> [`main`/examples/YOLOv8-OpenVINO-CPP-Inference/inference.cc](https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-OpenVINO-CPP-Inference/inference.cc)

I followed the instructions in this example. On line 27 of the inference.cc file, it gives a memory error when it tries to read the model.:
‍‍‍‍‍‍```
	std::shared_ptr<ov::Model> model = core.read_model(model_path); // Read the model from file
```

I have OpenVINO 2024.6.0, OpenCV 4.10 and Visual Studio 2022

**Comment by Y-T-G** [comment]
You can ignore the exception.

**Comment by nikpayam** [comment]
> You can ignore the exception.

I tried to handle the exception and ignore it, but the next line got another exception. Ignoring exceptions is not the right way. In the example you mentioned, OpenVINO cannot read the model file.

**Comment by Y-T-G** [comment]
I used to get that and it worked fine if you ignore it 

https://github.com/openvinotoolkit/openvino/issues/18371

**Comment by nikpayam** [comment]
> I used to get that and it worked fine if you ignore it
> 
> [openvinotoolkit/openvino#18371 (comment)](https://github.com/openvinotoolkit/openvino/issues/18371#issuecomment-1707863846)

I added a line to the end of the main function that writes something to the console. When I compile the app (in both release and debug modes) and run it in the console, the app runs but somewhere in the middle it exits and nothing is written to the console. How can I prevent the app from exiting when it encounters an exception?

**Comment by Y-T-G** [comment]
You can use try catch to ignore them depending on where they occur.

**Comment by BigMuscle85** [comment]
Have you solved this issue?
I have the similar issue when trying to detect objects using Yolo10 with OpenCV in C++. 

I just changed Detect module in Yolo11 to v10Detect to benchmark the performance. Doing the detection using the same trained model with Ultralytics works correctly thus the model is trained properly.
The output of such network is different. Original Detect module gives [1, 9, 525] output share, but v10Module gives [1, 300, 6]. According the code https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/modules/head.py#L97 the output channels are swapped for Yolo10, so I swap them back in my code:

```
Mat blob = dnn::blobFromImage(image, 1.0 / 255.0, Size(netInputWidth, netInputHeight), Scalar(), false, false);
net.setInput(blob);

// run inference
cv::Mat detection = net.forward();
cv::transposeND(detection, { 0, 2, 1 }, detection);

cv::Mat preds = detection.reshape(1, detection.size[1]); // [1, 8400, 85] -> [8400, 85]
for (int i = 0; i < preds.cols; i++) {
   float x = preds.at<float>(0, i);
   float y = preds.at<float>(1, i);
   float w = preds.at<float>(2, i);
   float h = preds.at<float>(3, i);
   float confidence = preds.at<float>(4, i);
   int class_id = static_cast<int>(preds.at<float>(5, i));
```

The problem is that produced data is not usable. All confidences are very close to zero, although class_id looks as a valid number in correct range. Maybe some other post-processing is required?

**Comment by Y-T-G** [comment]
You shouldn't be swapping for YOLOv10.

And the output is xyxy

**Comment by BigMuscle85** [comment]
But it does not make sense then, because output channels are in different order ([1, 300, 6]) than for Yolo11 ([1, 9, 525]). So swapping must be done either for Yolo10 or Yolo11 and then use correctly rows/cols and preds.at<float>(0, i) vs preds.at<float>(i, 0).

My example above uses preds.cols for number of detections, i.e. 300 for Yolo10 and 525 for Yolo11.

For example, OpenCV yolo_detector example does swapping for all yolo8,9,10 but it is not correct either:
https://github.com/opencv/opencv/blob/36a5176a5fdd46f607c1a0119d6703820a1513ea/samples/dnn/yolo_detector.cpp#L123

**Comment by Y-T-G** [comment]
Swapping is done for YOLO11, but not YOLOv10

**Comment by BigMuscle85** [comment]
You are right but practically it makes no difference. If you swap for YOLO11 then you index detections by rows, If you swap for YOLO10 then you index detections by columns. In both cases, YOLO10 returns invalid detections in OpenCV.

**Comment by Y-T-G** [comment]
If the detections are wrong, then it simply means your preprocessing is wrong.

**Comment by BigMuscle85** [comment]
I understand. So the question is whether the preprocessing is different for YOLO10 vs YOLO11? The used preprocessing works correctly for YOLO11. The detections get wrong just with changing the Detect layer to v10Detect. The preprocessing is still same.

**Comment by Y-T-G** [comment]
It's the same
